aptos-move/aptos-aggregator/src/aggregator_change_set.rs

```diff
@@ -0,0 +1,507 @@
+}
+
+impl AggregatorChange {
+    /// TODO: Implement this.
+    pub fn materialize_aggregator(
+        self,
+        _state_view: &dyn StateView,
+        _aggregator_id: &AggregatorID,
+    ) -> anyhow::Result<AggregatorChange, VMStatus> {
+        Ok(self)
+    }
+
+    /// Returns the result of applying the AggregatorState to the base value
+    /// Returns error if postcondition is not satisfied.
+    pub fn apply_aggregator_change_to(&self, base: u128) -> PartialVMResult<u128> {
+        match self.state {
+            AggregatorState::Data { value } => Ok(value),
+            AggregatorState::Delta { delta, .. } => {
+                // First, validate if the current delta operation can be applied to the base.
+                validate_history(base, self.max_value, &self.state)?;
+                Ok(addition_deltavalue(base, delta, self.max_value)?)
+            },
+        }
+    }
+
+    /// Applies self on top of previous delta, merging them together. Note
+    /// that the strict ordering here is crucial for catching overflows
+    /// correctly.
+    ///
+    /// TODO: What happens if the base aggregator is not none?
+    pub fn merge_with_previous_aggregator_change(
+        &mut self,
+        previous_change: AggregatorChange,
+    ) -> PartialVMResult<()> {
+        assert_eq!(
+            self.max_value, previous_change.max_value,
+            "Cannot merge aggregator changes with different max_values",
+        );
+        // When the previous aggregator change is a snapshot, we cannot merge it with the current.
+        if previous_change.base_aggregator.is_some() {
+            return Err(
+                PartialVMError::new(StatusCode::UNKNOWN_INVARIANT_VIOLATION_ERROR)
+                    .with_message("Cannot merge aggregator snapshots".to_string()),
+            );
+        }
+        match self.state {
+            AggregatorState::Data { .. } => {
+                // If the current state is Data, then merging with previous state won't change anything.
+                return Ok(());
+            },
+            AggregatorState::Delta { delta, history, .. } => {
+                match previous_change.state {
+                    AggregatorState::Data { value } => {
+                        // When prev_state is Data { value }, and current state is Delta { delta }, merging them into Data { value + delta }
+                        validate_history(value, self.max_value, &self.state)?;
+                        self.state = AggregatorState::Data {
+                            value: addition_deltavalue(value, delta, self.max_value)?,
+                        };
+                    },
+                    AggregatorState::Delta {
+                        speculative_start_value: _,
+                        delta: prev_delta,
+                        history: prev_history,
+                    } => {
+                        let new_delta = prev_delta.add(&delta, self.max_value)?;
```

Comment:
after the code below is re-structured, let's also do it in a way that we re-use code from here (everything is based on prev_delta, prev_history, self delta and self history & max value) to also do merging for DeltaOps? I'll add a comment there as well